rosbag2_compression/src/rosbag2_compression/sequential_compressor_reader.cpp

```diff
@@ -0,0 +1,195 @@
+// limitations under the License.
+
+#include "rosbag2_compression/sequential_compressor_reader.hpp"
+
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rcpputils/filesystem_helper.hpp"
+#include "rosbag2_compression/compression_options.hpp"
+#include "rosbag2_compression/zstd_decompressor.hpp"
+#include "logging.hpp"
+
+
+namespace rosbag2_compression
+{
+
+SequentialCompressorReader::SequentialCompressorReader(
+  std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory,
+  std::shared_ptr<rosbag2_cpp::SerializationFormatConverterFactoryInterface> converter_factory,
+  std::unique_ptr<rosbag2_storage::MetadataIo> metadata_io)
+: storage_factory_(std::move(storage_factory)),
+  converter_factory_(std::move(converter_factory)),
+  metadata_io_(std::move(metadata_io))
+{}
+
+SequentialCompressorReader::~SequentialCompressorReader()
+{
+  reset();
+}
+
+void SequentialCompressorReader::reset()
+{
+  storage_.reset();
+}
+
+void SequentialCompressorReader::open_storage()
+{
+  storage_ = storage_factory_->open_read_only(
+    *current_file_iterator_, metadata_.storage_identifier);
+  if (!storage_) {
+    throw std::runtime_error{"No storage could be initialized. Abort"};
+  }
+}
+
+void SequentialCompressorReader::setup_compression()
+{
+  compression_mode_ = rosbag2_compression::compression_mode_from_string(metadata_.compression_mode);
+  if (compression_mode_ != rosbag2_compression::CompressionMode::NONE) {
+    if (metadata_.compression_format == "zstd") {
+      decompressor_ = std::make_unique<rosbag2_compression::ZstdDecompressor>();
+    } else {
+      std::stringstream err;
+      err << "Unsupported compression format " << metadata_.compression_format;
+      throw std::invalid_argument{err.str()};
+    }
+  }
+}
+
+void SequentialCompressorReader::open(
+  const rosbag2_cpp::StorageOptions & storage_options,
+  const rosbag2_cpp::ConverterOptions & converter_options)
+{
+  // If there is a metadata.yaml file present, load it.
+  // If not, let's ask the storage with the given URI for its metadata.
+  // This is necessary for non ROS2 bags (aka ROS1 legacy bags).
+  if (metadata_io_->metadata_file_exists(storage_options.uri)) {
+    metadata_ = metadata_io_->read_metadata(storage_options.uri);
+    if (metadata_.relative_file_paths.empty()) {
+      ROSBAG2_COMPRESSION_LOG_WARN("No file paths were found in metadata.");
+      return;
+    }
+    open_storage();
+    setup_compression();
+    file_paths_ = metadata_.relative_file_paths;
+    current_file_iterator_ = file_paths_.begin();
+  } else {
+    open_storage();
+    metadata_ = storage_->get_metadata();
+    if (metadata_.relative_file_paths.empty()) {
+      ROSBAG2_COMPRESSION_LOG_WARN("No file paths were found in metadata.");
+      return;
+    }
+    file_paths_ = metadata_.relative_file_paths;
+    current_file_iterator_ = file_paths_.begin();
+  }
```

Comment:
You can just throw an exception here if `metadata_file_exists` returns false since ROS1 bag v2 files compressed with ZSTD are out of scope.